Bank login

KrasimirRaikov opened the pull request:
When the user logs in, the page writes the sessionId into the database with expiration time, and makes a cookie with the sessionId for the user. If the user is loged in correctly, he is redirected to the Home page, if not a message is displayed.

![login](https://cloud.githubusercontent.com/assets/13517204/16609034/5217f35a-4359-11e6-995a-ff5aa1eff994.png)

![homepage](https://cloud.githubusercontent.com/assets/13517204/16609040/5ecc45d8-4359-11e6-9267-17c8ed4167c3.png)

![login_wrongpassword](https://cloud.githubusercontent.com/assets/13517204/16609042/640b4d6e-4359-11e6-9c2e-231f32e18982.png)

panayotkulchev commented on src/main/java/com/clouway/bank/core/LoginRepository.java:
```diff
@@ -0,0 +1,16 @@
+ *
+ * @author Krasimir Raikov(raikov.krasimir@gmail.com)
+ */
+public interface LoginRepository {
```
SessionRepostory

KrasimirRaikov replied:
Done.

panayotkulchev commented on src/main/java/com/clouway/bank/core/LoginRepository.java:
```diff
@@ -0,0 +1,16 @@
+   *
+   * @param login the login to be added
+   */
+  void add(Login login);
```
creaste(session);

KrasimirRaikov replied:
done.

panayotkulchev commented on src/main/java/com/clouway/bank/core/Session.java:
```diff
@@ -0,0 +1,37 @@
+public class Session {
+  public final String id;
+  public final String userId;
+  public final Long expTime;
```
`expirationTime`

KrasimirRaikov replied:
done

panayotkulchev commented on src/main/java/com/clouway/bank/http/BankTimeCalendar.java:
```diff
@@ -0,0 +1,48 @@
+    calendar.setTime(new Date(System.currentTimeMillis()));
+    calendar.set(Calendar.MILLISECOND, 0);
+    calendar.set(Calendar.MINUTE, calendar.get(Calendar.MINUTE) + delayInMinutes);
+    return calendar.getTimeInMillis();
```
If you need millis of the current system time can just `System.currentTimeMillis()` and remove this method. Are you sure you will need all these Calendar thinks?

KrasimirRaikov replied:
I have made them, so I can easily replace the implementation and so that I don't have to mess with the system time in order to test, the classes that depend on that time.

panayotkulchev commented on src/main/java/com/clouway/bank/http/LoginController.java:
```diff
@@ -0,0 +1,98 @@
+    calendar = builder.calendar;
+  }
+
+  public static Builder newBuilder() {
```
dont you think using builder in this situation is strange

KrasimirRaikov replied:
Perhaps, but I have 4 parameters here.

panayotkulchev commented on src/main/java/com/clouway/bank/http/LoginController.java:
```diff
@@ -0,0 +1,58 @@
+      User user = userRepository.getUserById(username);
+
+      String message = userValidator.passwordsMatch(user.password, password);
+      if ("".equals(message)) {
```
Could you simplify this composite `if else` blocks

KrasimirRaikov replied:
I simplified it by using 
``` java
if(contidition){
some code  
return;}
``` 
blocks

panayotkulchev commented on src/main/java/com/clouway/bank/http/LoginFilter.java:
```diff
@@ -0,0 +1,68 @@
+      Session session = sessionRepository.retrieve(sessionCookie.getValue());
+      if (session.expirationTime < bankCalendar.getCurrentTime()) {
+        chain.doFilter(request, response);
+      } else {
```
`if else` and in the `else` you have new `if else` it is hard to read and understand

KrasimirRaikov replied:
I simplified it by using 
``` java
if(contidition){
some code  
return;}
``` 
blocks

panayotkulchev commented on src/main/java/com/clouway/bank/http/SecurityFilter.java:
```diff
@@ -0,0 +1,71 @@
+        sessionId.setMaxAge(0);
+        response.addCookie(sessionId);
+        response.sendRedirect("/login");
+      }else{
```
...

KrasimirRaikov replied:
I simplified it by using 
``` java
if(contidition){
some code  
return;}
``` 
blocks

panayotkulchev commented on src/main/java/com/clouway/bank/persistent/PersistentSessionRepository.java:
```diff
@@ -0,0 +1,92 @@
+      return new Session(resultSet.getString("sessionid"), resultSet.getString("username"), resultSet.getTimestamp("expirationtime").getTime());
+    } catch (SQLException e) {
+      e.printStackTrace();
+      throw new LoginException("There was a problem with your session, please try to login again");
```
When this error will be thrown. and if you throw this exception why you do not use it in the `update(session)` too?

KrasimirRaikov replied:
Ще се хвърли при грешка на взаймодействието с базата, може би трябва да го има и при ```update(session)```

mlesikov commented:
http://jobtiger.tv/hr-industriya/prognozata-na-svetlin-nakov-za-budeschteto-ot-it-tata-do-taksitata

panayotkulchev commented on src/main/java/com/clouway/bank/SessionCleaner.java:
```diff
@@ -0,0 +1,53 @@
+      public void run() {
+        String query = "DELETE  FROM login WHERE expirationtime < ?;";
+        try {
+          Connection connection = connectionManager.get();
```
What happens with the connection ofter job is done

KrasimirRaikov replied:
I've fixed that, I close it, when I'm finished with it.

panayotkulchev commented on src/main/java/com/clouway/bank/http/SecurityFilter.java:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  private Cookie getCookie(String cookieName, Cookie[] cookies) {
+    Cookie resultCookie = new Cookie("1", "");
```
Fix this

KrasimirRaikov replied:
I've made method which gives me my sessionId directly.

panayotkulchev commented on src/main/java/com/clouway/bank/http/SecurityFilter.java:
```diff
@@ -0,0 +1,83 @@
+public class SecurityFilter implements Filter {
+  private SessionRepository sessionRepository;
+  private BankCalendar bankCalendar;
+  private List<String> accessible;
```
something like `unsecuredPages`

KrasimirRaikov replied:
I've renamed it to ```unsecuredPages```

panayotkulchev commented on src/main/java/com/clouway/bank/http/SecurityFilter.java:
```diff
@@ -0,0 +1,83 @@
+    boolean sessionActive = session.isPresent()&&session.get().expirationTime > bankCalendar.getCurrentTime();
+
+    if (uriAccessible && sessionActive) {
+      response.sendRedirect("/");
```
in case `/somePage` is requested, and uri is accessible and has active session why client is send to `"/"` in case other page is requested?

KrasimirRaikov replied:
the ```accessible```s  are login and register, if the user has active session, is redirected to the home.

panayotkulchev commented on src/test/java/com/clouway/bank/persistent/PersistentSessionRepositoryTest.java:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  @Test
+  public void addLogin() {
```
You are testing `SESSION` repository, your tests should sound NOT like `addLogin` `updateLogin` but `createSession`, `updateSession`. If you have user repository will be `createUser` or `addUser`

panayotkulchev commented on src/test/java/com/clouway/bank/utils/LoginRepositoryUtility.java:
```diff
@@ -0,0 +1,26 @@
+package com.clouway.bank.utils;
```
If is used only for testing it could be in package `com.clouway.bank.test`

KrasimirRaikov replied:
it is in ```test/java``` directory.

panayotkulchev replied:
ok

KrasimirRaikov commented:
PR notes are fixed

panayotkulchev commented on src/main/java/com/clouway/bank/SessionCleaner.java:
```diff
@@ -0,0 +1,65 @@
+      }
+    };
+
+
```
Above 15 lines looks not formatted. Yesterday I told you to review your PR before mark mark it for review. I spend so much time to comment theese mistakes.

KrasimirRaikov replied:
I'm sorry, I thought that I went through all the files from this PR and formatted them, but I have missed one, from now on I will format the whole project.

panayotkulchev commented on src/test/java/com/clouway/bank/http/LoginControllerTest.java:
```diff
@@ -0,0 +1,105 @@
+/**
+ * @author Krasimir Raikov(raikov.krasimir@gmail.com)
+ */
+public class LoginControllerTest {
```
I think not all scenarios are covered with tests.

KrasimirRaikov replied:
I have added scenario about passwords not matching.

panayotkulchev commented:
After fix the 2 comments above you have LGTM

panayotkulchev commented:
After try to register with different passwords I was send to login with no error message. Fix it.
After login with use demo and pass demo some error message with null was displayed. Fix it.
Please test all scenarios. And when you are shure everything is redy notify me.

KrasimirRaikov commented:
I have fixed the problem with displaying message when passwords don't match on registration, and displaying the message when the user is not found on login. Also fixed a problem with the displaying of message when username is taken during registration, and displaying the message when there is insufficient amount in the account for withdraw.

panayotkulchev commented:
You commit messages are like: 
![commitmessageskrasi](https://cloud.githubusercontent.com/assets/8696264/16829924/687febf6-49a4-11e6-8472-702d4c5e08d0.jpg)

Can be like:

Bank login: login page
Bank login: login controller
Bank login: security filter
Bank login: refactoring tests
Bank login: fixed bug at user login when user do not exists in DB
....

panayotkulchev commented:
LGTM!
